### PR TITLE
[examples] nanoChatGPT

### PR DESCRIPTION
**Pull request by tcbegley**
## Description

Example recreation of a chatGPT-like model built on nanoGPT, TorchRL and tensordict.

### CODE REVIEW

File: examples/nanoChatGPT/train_rlhf.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    # DataLoader
+    train_loader, _ = get_dataloaders(config)
+
+    # Environment
+    env = RLHFEnv(reward_model=reward_model, config=config, dataloader=train_loader)
+
+    # ######## TRAINING LOOP ########
+
+    def get_action(td):
+        critic(td)
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:47]: no need for a critic in a policy

File: examples/nanoChatGPT/train_rlhf.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    for i in range(config["max_iters"]):
+        td = env.rollout(
+            config["episode_length"], policy=get_action, return_contiguous=False
+        )
+
+        # TODO: add replay buffer?
+        with set_skip_existing(True):
+            adv_fn(td)
+            loss_vals = loss_fn(td)
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:60]: putting loss under skip_existing is dangerous. 
The best is to pass the actor and *just* the critic head to the ppo loss (not the full critic).

Comment (vmoens) [reply examples/nanoChatGPT/train_rlhf.py:60]: See for instance `get_value_head` here https://pytorch.org/rl/reference/generated/torchrl.modules.tensordict_module.ActorCriticOperator.html#torchrl.modules.tensordict_module.ActorCriticOperator

File: examples/nanoChatGPT/train_rlhf.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+        loss_val = sum(
+            value for key, value in loss_vals.items() if key.startswith("loss")
+        )
+        loss_val.backward()
+        optimizer.step()
+        optimizer.zero_grad()
+
+        # Logging
+        print(f"Iteration {i}: {loss_val=}")
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:70]: I believe this is empty

File: examples/nanoChatGPT/models/actor_critic.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    model_base, _ = init_transformer(
+        config, as_tensordictmodule=False, skip_compilation=True
+    )
+    a2c_model = ActorCritic(model_base)
+    a2c_model.to(config["device"])
+    actor = a2c_model.get_policy_operator()
+    critic = a2c_model.get_value_operator()
+
+    # FIXME: we are missing compile...
+    # but we would compile TDModule...check performance issues
```
Comment (vmoens) [inline examples/nanoChatGPT/models/actor_critic.py:62]: hint: to compile a TDModule, we can use the Tracer that Tom coded, and then use torch.compile on the traced model (inside the traced TDModule, so to say).
We need a better way of doing this...

**Review by vmoens — COMMENTED** [review]
Some quick comments
